### DISCUSSION TITLE
How Minting Works

### DISCUSSION THREAD

**Discussion by allwin199** [Cyfrin/foundry-full-course-cu#515] [Q&A] [answered]
Whenever we call `_safeMint()` it mints the NFT using tokenId and tokenURI.

But I am finding it hard to understand, how `tokenURI` is called during `_safeMint()`

I want to understand, what happens behind the scenes.

**Comment by Niferu** [comment]
Hi @allwin199 

It is not. We have coded it by ourselves to assign correct tokenURI per created tokenId.

It is this line in our mint function:
```solidity
 s_tokenIdToUri[s_tokenCounter] = tokenUri;
```

After that we have overwritten `tokenURI()` function from ERC721 to return correct tokenURI mapped to correct tokenId

**Reply by allwin199** [reply]
https://github.com/Cyfrin/foundry-nft-f23/blob/fd791c5312fffa20af7a139093e9b7cf2d23cc73/src/MoodNft.sol#L55

In `mintNft()` tokenURI is not defined.
can you explain how it is working for this?

**Reply by Niferu** [reply]
Because here our uri is not specified as there are 2 posibilities and we are using baseURI, which is automatically assigned in original ERC721 code of tokenURI function, so we do not have to override it. We do override it because we have 2 posibilities and thats why. Otherwise using baseURI is already coded in original ERC721.

I highly recommend you reviewing carefully code of original ERC721 to get better understanding of using it in our projects.

**Reply by allwin199** [reply]
@Neftyr I get your point. can you show where it defined that, take baseURI as default.
I am searching for it and couldn't find.

```
/**
     * @dev See {IERC721Metadata-tokenURI}.
     */
    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
        _requireMinted(tokenId);

        string memory baseURI = _baseURI();
        return bytes(baseURI).length > 0 ? string(abi.encodePacked(baseURI, tokenId.toString())) : "";
    }

    /**
     * @dev Base URI for computing {tokenURI}. If set, the resulting URI for each
     * token will be the concatenation of the `baseURI` and the `tokenId`. Empty
     * by default, can be overridden in child contracts.
     */
    function _baseURI() internal view virtual returns (string memory) {
        return "";
    }
   ```
   
   This is the code from ERC721 I am referring to

**Reply by Niferu** [reply]
You have already linked exact part of code refering to that. This is only template shown in ERC721 as you can see. We can do something like here [link](https://github.com/PatrickAlphaC/hardhat-nft-fcc/blob/main/contracts/BasicNft.sol) and instead of `TOKEN_URI` we can name it baseURI as it is the same. As you can see we are not specifying tokenURI in mint function as it is defined as baseURI (one URI for all tokens minted or veery similar URI with same link but only last letter of that link, which refers to tokenId number changes). For example if our base tokenURI is something like: "ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json" we are changing it's number only ?filename=1-PUG.json, ?filename=2-PUG.json etc. but in this case we need to correct tokenURI() function to assign correct id and modify our baseURI.

I can also add that I have finished both python and JS modes of this course and I thought I understand NFT's, but once I started to work on my project based on NFT for hackathon I figured out that I know nothing and dont understand a thing from course, so I highly recommend you preparing your own NFT project maybe not so big as for hackathon but just to understand idea of tokenURI's, metadata, img links that pointing to image etc. By doing it from scratch on your own and testing blockchain behaviour by releasing some NFT's you will understand it all :)

It is not so easy topic as it might look like.

**Reply by allwin199** [reply]
gosh! This is complex.

Can you show an example of your project where you created an NFT?

**Reply by Niferu** [reply]
Sure take a look at below:

https://github.com/Neftyr/Digital-Rights-Maykr
https://github.com/MarkizOrias/ChainlinkHackathon2023

**Reply by allwin199** [reply]
Thanks. I will look into it.

**Reply by Niferu** [reply, marked as answer]
To simplify this `tokenURI` coming from ERC721 is responsible for pointing to correct metadata uri per tokenId minted. Metadata from that URI is pointing to correct image link. I hope I have answered your question

**Reply by allwin199** [reply]
yup! thanks

**Reply by Ynyesto** [reply]
Hi! Just wanted to ask if you'd know why it could be that all I get are empty images. I believe I did everything right and understand perfectly well the explanation, but when I import in Metamask, all I get is an empty image. [Here](https://sepolia.etherscan.io/address/0xb9D53d659Cf041D7bA7662E0607BE5AAd373Bb32#code)'s the contract of the BasicNft I deployed during lesson 11, with which I believe there's nothing wrong. However, as you can see, I get no image:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/33223441/a632584a-9750-48f5-8db7-c6fc621a39c2)
The same thing happens to me when I do the deployment on Anvil later in the lesson:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/33223441/25dff762-dc71-4cae-ae6a-86b5837e30b3)
I've checked the token URI for both cases (IPFS and SVG-base64) and have no clue what could be wrong...
Any ideas? Also, please let me know if I should make a separate issue of this. Thanks beforehand!

**Reply by Niferu** [reply]
Hi @arynyestos 

There was already couple of discussions already answered with same issue. Search for those and you will get an answer.

**Reply by Ynyesto** [reply]
Hmm, I did search, and this thread is the closest I got 🤔🤔🤔 I suppose I'll need to take another look. Will let you know if I can't find it. Thanks!

**Reply by Niferu** [reply]
@arynyestos 

Check JS course discussions also
